Adds new tutorial for Nested SPIRE deployments (#25)
spiffe/spire-tutorials: nested-spire/README.md

```console
   services:
     # Root
     root-server:
       image: gcr.io/spiffe-io/spire-server:0.10.1
       hostname: root-server
       volumes:
         - ./root/server:/opt/spire/conf/server
       command: ["-config", "/opt/spire/conf/server/server.conf"]
     root-agent:
       # Share the host pid namespace so this agent can attest the nested servers
       pid: "host"
       image: gcr.io/spiffe-io/spire-agent:0.10.1
       depends_on: ["root-server"]
       hostname: root-agent
       volumes:
         # Share root agent socket to be acceded by nestedA and nestedB servers
         - ./sharedRootSocket:/opt/spire/sockets
         - ./root/agent:/opt/spire/conf/agent
         - /var/run/:/var/run/
       command: ["-config", "/opt/spire/conf/agent/agent.conf"]
```

## Configuring NestedA SPIRE Deployment

> The same set of configurations are required for the NestedB SPIRE deployment but those changes are not described in the text to avoid needless repetition.

Reviewer: Ok, we can leave the new sentence out.

> The same set of configurations are required for the NestedB SPIRE deployment but those changes are not described in the text to avoid needless repetition.

Author: Perfect. I'll make the change and mark this comment as resolved. Thanks.

> SPIRE Server can be configured using different types of plugins. For Nested SPIRE deployments we use the UpstreamAuthority type that allows SPIRE server to integrate with existing PKI systems. For the guide we use the `spire` UpstreamAuthority plugin which uses an upstream SPIRE server in the same trust domain to obtain intermediate signing certificates for SPIRE server.

Reviewer: Ok, here are some adjustments to your new text. To me, it's a little confusing that UpstreamAuthority is a "type" that includes 5 different plugins. So I've tried to clarify that here with a new sentence.--

SPIRE Agent and Server can be extended with a variety of [plugins](https://spiffe.io/spire/docs/extending/). The [UpstreamAuthority plugin](https://github.com/spiffe/spire/blob/master/doc/spire_server.md#built-in-plugins) type allows SPIRE Server to integrate with existing PKI systems. UpstreamAuthority plugins can sign certificates using CAs loaded from disk, third-party tools from AWS and Vault, and so on. Nested SPIRE deployments require the use of the [spire UpstreamAuthority plugin](https://github.com/spiffe/spire/blob/master/doc/plugin_server_upstreamauthority_spire.md) which uses an upstream SPIRE Server in the same trust domain to obtain intermediate signing certificates for SPIRE Server.

> SPIRE Server can be configured using different types of plugins. For Nested SPIRE deployments we use the UpstreamAuthority type that allows SPIRE server to integrate with existing PKI systems. For the guide we use the `spire` UpstreamAuthority plugin which uses an upstream SPIRE server in the same trust domain to obtain intermediate signing certificates for SPIRE server.

Author: I like it. I'll use this new version. Thanks Steve.

The other point to highlight is the `-downstream` option. This option when set, indicates that the entry describes a downstream SPIRE server.

## Run the Scenario

Use the `set-env.sh` script to run all the services that compound the scenario. All the configurations described in the previous sections are illustrated in the different services.
Ensure that the current working directory is `.../spire-tutorials/nested-spire` and run:

```console
    bash scripts/set-env.sh
```

Once the script is completed, in another terminal run the following command to review the logs from all the services

```console
    docker-compose logs -f -t
```

# Part 2: Test the Deployments

Now that the SPIRE deployments are ready, let's test the scenario that we've configured.

## Create Workload Registration Entries

> To test the scenario we create two workload registration entries, one entry per each of the nested SPIRE deployment. The goal of the test case is to demonstrate that SVIDs created in a nested configuration are valid in the entire trust domain, not only in the scope of the server that originated the SVID.

Reviewer: You used "per" correctly here, but "for" just sounds slightly better.

```suggestion
To test the scenario we create two workload registration entries, one entry for each nested SPIRE Server (`nestedA` and `nestedB`). The goal of the test is to demonstrate that SVIDs created in a nested configuration are valid in the entire trust domain, not only in the scope of the SPIRE Server that originated the SVID. The following commands demonstrate the command line options we'll use to create the two workload registration entries, but you can run these commands using the `create-workload-registration-entries.sh` script shown a few lines below.
```

```console
   # Workload for nestedA deployment
   docker-compose exec -T nestedA-server \
       /opt/spire/bin/spire-server entry create \
       -parentID "spiffe://example.org/spire/agent/x509pop/$(fingerprint nestedA/agent/agent.crt.pem)" \
       -spiffeID "spiffe://example.org/nestedA/workload" \
       -selector "unix:uid:1001" \
       -ttl 0

   # Workload for nestedB deployment
   docker-compose exec -T nestedB-server \
       /opt/spire/bin/spire-server entry create \
       -parentID "spiffe://example.org/spire/agent/x509pop/$(fingerprint nestedB/agent/agent.crt.pem)" \
       -spiffeID "spiffe://example.org/nestedB/workload" \
       -selector "unix:uid:1001" \
       -ttl 0
```

> The examples uses the `$(fingerprint nestedA/agent/agent.crt.pem)` notation again to show that the `-parentID` flag is set to the SPIFFE ID of the nested SPIRE Agent. TTL equals to zero indicates that the default value of 3600 seconds will be used. Finally, in both cases there is an unix selector to identify any process that is run by the user with id 1001.

Reviewer: Since UNIX is pronounced "yoo-nix" it's "a unix" selector. But if you call for a car using an app, it's "getting an Uber" (oo-brr).

Please check my changes for any wrong information.

```suggestion
The examples use the `$(fingerprint nestedA/agent/agent.crt.pem)` form again to show that the `-parentID` flag specifies the SPIFFE ID of the nested SPIRE Agent. The TTL flag value of zero indicates that the default value of 3600 seconds will be used. Finally, in both cases the unix selector assigns the SPIFFE ID to any process with a uid of 1001.
```

> Use the following Bash script to create the registration entries:

Reviewer: ```suggestion
Use the following Bash script to create the registration entries using the options just described:
```

```console
   bash scripts/create-workload-registration-entries.sh
```

## Run the Test

> Once both workload registration entries are propagated, let's test that SVIDs created in a nested configuration are valid in the entire trust domain, not only in the scope of the server that originated the SVID.

Reviewer: ```suggestion
Once both workload registration entries are propagated, let's test that SVIDs created in a nested configuration are valid in the entire trust domain, not only in the scope of the SPIRE Server that originated the SVID.
```

> The test consists on getting a JWT-SVID from the `nestedA-agent` and validate it using the `nestedB-agent`. In both cases, Docker Compose run the processes using the user 1001 to match the workload registration entries created in the previous section.

Reviewer: ```suggestion
The test consists of getting a JWT-SVID from the `nestedA-agent` SPIRE Agent and validating it using the `nestedB-agent`. In both cases, Docker Compose runs the processes using the uid 1001 to match the workload registration entries created in the previous section.
```

> 

Reviewer: I think we can run these commands with the test.sh script, right?

```suggestion
The `test.sh` script runs the following commands:
```

Author: Yes, test.sh run the same commands to check the configuration is correct.
I didn't use the script intentionally to force the reader to run it manually as part of the tutorial the same way it is done with the other tutorials in the repository.
I think test.sh is something internal and because of that I never mention it.

That said, I'm open to change it if you consider it a better approach.

Reviewer: That's an excellent point, I agree with you. We shouldn't use test.sh but instead show the actual commands.

Author: Good. I'll discard this suggestion and also the next one that is linked to this comment. 
I'll mark both as resolved.

```console
    # Fetch JWT-SVID and extract token
    token=$(docker-compose exec -u 1001 -T nestedA-agent \
      /opt/spire/bin/spire-agent api fetch jwt -audience nested-test -socketPath /opt/spire/sockets/workload_api.sock | sed -n '2p')

    # Validate token
    docker-compose exec -u 1001 -T nestedB-agent \
        /opt/spire/bin/spire-agent api validate jwt -audience nested-test  -svid "${token}" \
          -socketPath /opt/spire/sockets/workload_api.sock
```

> 

Reviewer: ```suggestion
To run the `test.sh` script which fetches and validates the token, ensure that the current working directory is `.../spire-tutorials/nested-spire` and type:

 ```console
     bash scripts/test.sh
 ```
```

The result indicates that the JWT-SVID is valid and shows the SPIFFE ID associated to the JWT-SVID which belongs to the workload registered on the `nestedA-agent`. 

```console
    SVID is valid.
    SPIFFE ID : spiffe://example.org/nestedA/workload
    Claims    : {"aud":["nested-test"],"exp":1595814190,"iat":1595813890,"sub":"spiffe://example.org/nestedA/workload"}
```

In SPIRE this is accomplished by propagating every JWT-SVID public signing key to the whole topology. In the case of X509-SVID, this is easily achieved because of the chaining semantics that X.509 has.

# Cleanup

> When you are finished running this tutorial, you can use the following command to stop all the containers

Reviewer: ```suggestion
When you are finished running this tutorial, you can use the following command to stop all the containers:
```

Use the following Bash script to create the registration entries using the options just described:

```console
   bash scripts/create-workload-registration-entries.sh
```

## Run the Test

Once both workload registration entries are propagated, let's test that SVIDs created in a nested configuration are valid in the entire trust domain, not only in the scope of the SPIRE Server that originated the SVID.

The test consists of getting a JWT-SVID from the `nestedA-agent` SPIRE Agent and validating it using the `nestedB-agent`. In both cases, Docker Compose runs the processes using the uid 1001 to match the workload registration entries created in the previous section.

```console
    # Fetch JWT-SVID and extract token
    token=$(docker-compose exec -u 1001 -T nestedA-agent \
      /opt/spire/bin/spire-agent api fetch jwt -audience nested-test -socketPath /opt/spire/sockets/workload_api.sock | sed -n '2p')

    # Validate token
    docker-compose exec -u 1001 -T nestedB-agent \
        /opt/spire/bin/spire-agent api validate jwt -audience nested-test  -svid "${token}" \
          -socketPath /opt/spire/sockets/workload_api.sock
```

> The result indicates that the JWT-SVID is valid and shows the SPIFFE ID associated to the JWT-SVID which belongs to the workload registered on the `nestedA-agent`. 

Reviewer: ```suggestion
The `nestedB` SPIRE Agent outputs the following:

<move output here>

The output indicates that the JWT-SVID is valid. Also, although the SPIFFE ID is registered on `nestedA` rather than `nestedB`, the SPIFFE ID is still valid on the `nestedB` SPIRE Agent because the SPIRE Agents are in the same trust domain in the nested SPIRE topology. 
```

# Overview

Nested SPIRE allows SPIRE Servers to be “chained” together, and for all SPIRE Servers to issue identities in the same trust domain, meaning all workloads identified in the same trust domain are issued identity documents that can be verified against the root keys of the trust domain.

Nested topologies work by co-locating a SPIRE Agent with every downstream SPIRE Server being “chained”. The downstream SPIRE Server obtains credentials over the Workload API that it uses to directly authenticate with the upstream SPIRE Server to obtain an intermediate CA.

To demonstrate a deployment of SPIRE in a nested topology we create a scenario using Docker Compose with a root SPIRE deployment and two SPIRE deployments nested to it.

![Nested SPIRE diagram][nested-SPIRE-diagram]

> [nested-SPIRE-diagram]: images/Nested_SPIRE_Diagram.png "Nested SPIRE Deployment diagram"

Reviewer 2: this diagram is not right, it is true that nestedA uses agent but it uses agent in order to fetch an svid for itself and then uses that svid to communicate with root-server in order to push mints, maybe you can udpate that diagram to reflect that?

```console
   UpstreamAuthority "spire" {
 	   plugin_data = {
 	       server_address      = "root-server"
 	       server_port         = 8081
 	       workload_api_socket = "/opt/spire/sockets/workload_api.sock"
 	   }
    }
```

The Docker Compose definition for the `nestedA-server` service in the [docker-compose.yaml](docker-compose.yaml) file mounts the new local directory `sharedRootSocket` as a volume. Remember from the previous section that the `root-agent` socket is mounted on that directory. That way the `nestedA-server` can access the `root-agent` workload API and fetch its SVID.

```console
   nestedA-server:
     # Share the host pid namespace so this server can be attested by the root agent
     pid: "host"
     image: gcr.io/spiffe-io/spire-server:0.10.1
     hostname: nestedA-server
     labels:
       # label to attest nestedA-server against root-agent
       - org.example.name=nestedA
     volumes:
       # Add root agent socket  
       - ./shared/rootSocket:/opt/spire/sockets
       - ./nestedA/server:/opt/spire/conf/server
     command: ["-config", "/opt/spire/conf/server/server.conf"]
```

## Create Downstream Registration Entry

> The `nestedA-server` must be registered on the `root-server` to get SVIDs. We achieve this by creating a registration entry in the root SPIRE Server for the `nestedA-server`.

Reviewer 2: why it is created? can you explain that this done in order to provide an identity to `nestedA-server` that will be used by plugin in order to mint svids?

Reviewer 2: and that `-spiffeID "spiffe://example.org/nestedA"` is the server's id

```console
   docker-compose exec -T root-server \
       /opt/spire/bin/spire-server entry create \
       -parentID "spiffe://example.org/spire/agent/x509pop/$(fingerprint root/agent/agent.crt.pem)" \
       -spiffeID "spiffe://example.org/nestedA" \
       -selector "docker:label:org.example.name:nestedA-server" \
       -downstream \

>        -ttl 3600

Reviewer 2: maybe for example purpose you can reduce that time? just to display people how downstream expires and is renewed?

```

The `-parentID` flag contains the SPIFFE ID of the `root-agent`. The SPIFFE ID of the `root-agent` is created by the [x509pop Node Attestor](https://github.com/spiffe/spire/blob/master/doc/plugin_server_nodeattestor_x509pop.md) plugin which defines the SPIFFE ID as `spiffe://<trust domain>/spire/agent/x509pop/<fingerprint>`. A `fingerprint()` function in the shell script calculates the SHA1 fingerprint of the certificate.
The other point to highlight is the `-downstream` option. This option, when set, indicates that the entry describes a downstream SPIRE Server.

## Run the Scenario

Use the `set-env.sh` script to run all the services that make up the scenario. The script starts the `root`, `nestedA`, and `nestedB` services with the configuration options described earlier.

> Ensure that the current working directory is `.../spire-tutorials/nested-spire` and run:

Reviewer 2: mmm can it be confuse to some point, in other case you used `    cd your_path/nested-spire` you must use the same in both cases

Reviewer: I agree with Marcos that we should use the same form in all places. I have a slight preference for your_path/spire-tutorials/nested-spire but .../spire-tutorials/nested-spire is ok also. (Of course, use just your_path/spire-tutorials in the earlier instance to show the precise appropriate path.)

Reviewer: I agree with Marcos that we should use the same form in all places. I have a slight preference for your_path/spire-tutorials/nested-spire but .../spire-tutorials/nested-spire is ok also. (Of course, use just your_path/spire-tutorials in the earlier instance to show the precise appropriate path.)

```console
   # Workload for nestedA deployment
   docker-compose exec -T nestedA-server \
       /opt/spire/bin/spire-server entry create \
       -parentID "spiffe://example.org/spire/agent/x509pop/$(fingerprint nestedA/agent/agent.crt.pem)" \
       -spiffeID "spiffe://example.org/nestedA/workload" \
       -selector "unix:uid:1001" \
       -ttl 0

   # Workload for nestedB deployment
   docker-compose exec -T nestedB-server \
       /opt/spire/bin/spire-server entry create \
       -parentID "spiffe://example.org/spire/agent/x509pop/$(fingerprint nestedB/agent/agent.crt.pem)" \
       -spiffeID "spiffe://example.org/nestedB/workload" \
       -selector "unix:uid:1001" \
       -ttl 0
```

> The examples use the `$(fingerprint nestedA/agent/agent.crt.pem)` form again to show that the `-parentID` flag specifies the SPIFFE ID of the nested SPIRE Agent. The TTL flag value of zero indicates that the default value of 3600 seconds will be used. Finally, in both cases the unix selector assigns the SPIFFE ID to any process with a uid of 1001.

Reviewer 2: it can be read than we use the same figerprint for both that is not true, maybe you can clarify it in another way?

Reviewer: Andrés' wording may not be too confusing to users, but here are some ideas:
"examples use the $(fingerprint nested[A|B]/agent/agent.crt.pem) form again"  ?
OR 
"examples use the fingerprint() function to substitute the certificate fingerprint part of the nestedA and nestedB SPIFFE IDs when setting those SPIFFE IDs as the parent ID of each workload"  ?

Author: What about this?
_The examples use the `fingerprint path/to/nested-agent-cert` form again to show that..._

The idea behind the `fingerprint` function is mentioned before in the tutorial, there is no need to repeat that. The goal of this paragraph is to highlight the use of the nested agents as parent id of the new workload entries.

Use the following Bash script to create the registration entries using the options just described:

```console
   bash scripts/create-workload-registration-entries.sh
```

## Run the Test

> Once both workload registration entries are propagated, let's test that SVIDs created in a nested configuration are valid in the entire trust domain, not only in the scope of the SPIRE Server that originated the SVID.

Reviewer 2: how can they verify that workload registration entries are propagated?

Author: The output of the bash script shows that the entries are propagated, there is no extra step here for the reader.

Reviewer 2: maybe you can add a simple way to verify it on readme? for people that dont want to run bash, and they can verify it esasly?

Type this command to fetch the JWT-SVID on the `nestedA` SPIRE Agent and extract the token from the JWT-SVID:

```console
    token=$(docker-compose exec -u 1001 -T nestedA-agent \
      /opt/spire/bin/spire-agent api fetch jwt -audience nested-test -socketPath /opt/spire/sockets/workload_api.sock | sed -n '2p')
```

Run the following command to validate the token from `nestedA` on the `nestedB` SPIRE Agent:

```console
    docker-compose exec -u 1001 -T nestedB-agent \
        /opt/spire/bin/spire-agent api validate jwt -audience nested-test  -svid "${token}" \
          -socketPath /opt/spire/sockets/workload_api.sock
```

The `nestedB` SPIRE Agent outputs the following:

```console
    SVID is valid.
    SPIFFE ID : spiffe://example.org/nestedA/workload
    Claims    : {"aud":["nested-test"],"exp":1595814190,"iat":1595813890,"sub":"spiffe://example.org/nestedA/workload"}
```

The output indicates that the JWT-SVID is valid. Also, although the SPIFFE ID is registered on `nestedA` rather than `nestedB`, the SPIFFE ID is still valid on the `nestedB` SPIRE Agent because the SPIRE Agents are in the same trust domain in the nested SPIRE topology.

> In SPIRE this is accomplished by propagating every JWT-SVID public signing key to the whole topology. In the case of X509-SVID, this is easily achieved because of the chaining semantics that X.509 has.

Reviewer 2: can you add steps to validate x509 svids as well? basically you can use openssl to do that

# Cleanup

> When you are finished running this tutorial, you can use the following command to stop all the containers:

Reviewer 2: can you add a script stop compose? I understand it is just a line but to be consistent with the fact we started on a bash too

Author: You mean to create a bash with the instruction to stop docker-compose and use that bash instead?
Like:
```
bash scripts/clean-env.sh
```

And `clean-env.sh` content is:
```
docker-compose down
```
_`clean-env.sh` is the same name used in the other tutorials._

Reviewer 2: yes and remove all temporal files/directories
